packages/starknet-snap/src/utils/superstruct.ts

```diff
@@ -184,3 +192,174 @@ export const createStructWithAdditionalProperties = (
+  classHash: optional(string()),
+  casm: optional(CompiledSierraCasmStruct),
+  compiledClassHash: optional(string()),
+});
+
+export const DeployAccountContractStruct = object({
+  classHash: nonempty(string()),
+  constructorCalldata: optional(RawArgsStruct),
+  addressSalt: optional(NumberStringStruct),
+  contractAddress: optional(string()),
+});
+
+export const UniversalDeployerContractPayloadStruct = object({
+  classHash: NumberStringStruct,
+  salt: optional(string()),
+  unique: optional(boolean()),
+  constructorCalldata: optional(RawArgsStruct),
+});
+
+export const BaseInvocationStruct = object({
+  // lets not accept optaional payload to reduce the complexity of the struct
+  // as the snap control the input
+  payload: unknown(),
+  type: enums([
+    TransactionType.DECLARE,
+    TransactionType.DEPLOY,
+    TransactionType.DEPLOY_ACCOUNT,
+    TransactionType.INVOKE,
+  ]),
+});
+
+export const InvocationsStruct = define<Invocations>(
+  // We do use a custom `define` for this type to avoid having to use a `union` since error
+  // messages are a bit confusing.
+  //
+  // Doing manual validation allows us to use the "concrete" type of each supported acounts giving
+  // use a much nicer message from `superstruct`.
+  'InvocationsStruct',
+  (value: unknown[]) => {
+    for (const invocation of value as Invocations) {
+      // This will also raise if `value` does not match any of the supported TransactionType!
+      const maskedInvocation = mask(invocation, BaseInvocationStruct);
+
+      const isArray = Array.isArray(maskedInvocation.payload);
+
+      let struct: Struct;
+
+      switch (invocation.type) {
+        case TransactionType.DECLARE:
+          struct = DeclareContractPayloadStruct;
+          if (isArray) {
+            throw new Error('Declare payload does not accept mutiple items');
+          }
+          break;
+        case TransactionType.DEPLOY:
+          struct = UniversalDeployerContractPayloadStruct;
+          break;
+        case TransactionType.DEPLOY_ACCOUNT:
+          struct = DeployAccountContractStruct;
+          if (isArray) {
+            throw new Error(
+              'Deploy account payload does not accept mutiple items',
+            );
+          }
+          break;
+        case TransactionType.INVOKE:
+          struct = CallDataStruct;
+          break;
+        default:
+          throw new Error('Invalid transaction type');
+      }
+
+      const [error] = validate(
+        maskedInvocation.payload,
+        isArray ? array(struct) : struct,
+      );
+
+      if (error !== undefined) {
+        return error;
+      }
+    }
+    return true;
+  },
+);
+
+export const UniversalDetailsStruct = assign(
+  V3TransactionDetailStruct,
+  object({
+    blockIdentifier: optional(string()),
+    maxFee: optional(NumberStringStruct),
+    skipValidate: optional(boolean()),
+    version: optional(TxVersionStruct),
+  }),
+);
+
+export const DeployAccountPayloadStruct = object({
```

Comment:
why you need this, if we dont need to verify, if we just need the type , then just declare the type in type file